src/compiler/checker.ts

```diff
@@ -14212,12 +14222,26 @@ namespace ts {
             return undefined;
         }
 
+        function isDiscriminantType(type: Type): boolean {
+            if (type.flags & TypeFlags.Union) {
+                if (type.flags & (TypeFlags.Boolean | TypeFlags.EnumLiteral)) {
+                    return true;
+                }
+                let combined = 0;
+                for (const t of (<UnionType>type).types) combined |= t.flags;
+                if (combined & TypeFlags.Unit && !(combined & TypeFlags.Instantiable)) {

```

Comment:
what happens if you have two two non-unit types? Should that still be discriminable?

Reply:
I'm pretty sure this change isn't supposed to be here, since it's https://github.com/Microsoft/TypeScript/pull/27695